File: ABC/301-350/ABC-317/ABC317-C.cpp
#include <iostream>
#include <vector>

class Edge {
public:
    int start, end, leng;
    Edge(int start, int end, int leng) : start(start), end(end), leng(leng) {}
};

int dfs(int node, std::vector<bool>& visited, std::vector<std::vector<Edge>>& graph) {
    visited[node] = true;
    int max_length = 0;
    
    for (const Edge& edge : graph[node]) {
        if (!visited[edge.end]) {
            max_length = std::max(max_length, edge.leng + dfs(edge.end, visited, graph));
        }
    }
    
    visited[node] = false;
    return max_length;
}

const int INF = 1e9;

int main() {
    int N, M;
    std::cin >> N >> M;
    
    std::vector<std::vector<Edge>> graph(N);
    for (int i = 0; i < M; ++i) {
        int u, v, w;
        std::cin >> u >> v >> w;
        graph[u - 1].emplace_back(u - 1, v - 1, w);
        graph[v - 1].emplace_back(v - 1, u - 1, w);
    }

    int max_length = 0;
    for (int i = 0; i < N; ++i) {
        std::vector<bool> visited(N, false);
        max_length = std::max(max_length, dfs(i, visited, graph));
    }

    std::cout << max_length << std::endl;
    
    return 0;
}

File: ABC/301-350/ABC-317/ABC317-C.py
# 普通のpythonでは通らないので注意
# (PyPyなら通る)

class Edge:
    def __init__(self, start, end, leng):
        self.start = start
        self.end = end
        self.leng = leng

def dfs(node, visited, graph):
    visited[node] = True
    max_length = 0
    
    for edge in graph[node]:
        if not visited[edge.end]:
            max_length = max(max_length, edge.leng + dfs(edge.end, visited, graph))
    
    visited[node] = False
    return max_length

INF = 10**9 

N, M = map(int, input().split())
graph = [[] for _ in range(N)]

for i in range(M):
    u, v, w = map(int, input().split())
    graph[u - 1].append(Edge(u - 1, v - 1, w))
    graph[v - 1].append(Edge(v - 1, u - 1, w))

max_length = 0
for i in range(N):
    visited = [False] * N
    max_length = max(max_length, dfs(i, visited, graph))

print(max_length)